### PR TITLE
Give a good error message when forking after torch.cuda.device_count()

### LINKED ISSUE #15734: PyTorch doesn't report useful error when forking after CUDA initialization (originally: Unknown cuda error with multiprocessing)
## 🐛 Bug

Strange cuda error when torch.cuda.device_count called before torch.cuda.get_device_properties with multiprocessing.

## To Reproduce

~~~~
import multiprocessing, torch

def mp_worker(gpu):
    print(torch.cuda.get_device_properties(gpu))

gpus = list(range(torch.cuda.device_count()))

processes = [multiprocessing.Process(target=mp_worker, args=(gpui,)) for gpui in gpus]
for process in processes:
    process.start()
for process in processes:
    process.join()
~~~~

causes this (unhelpful) error to occur: 
~~~~
  File "/usr/local/lib/python3.6/dist-packages/torch/cuda/__init__.py", line 144, in init
    _lazy_init()
  File "/usr/local/lib/python3.6/dist-packages/torch/cuda/__init__.py", line 162, in _lazy_init
    torch._C._cuda_init()
RuntimeError: cuda runtime error (30) : unknown error at /pytorch/aten/src/THC/THCGeneral.cpp:51
~~~~

If I change the code so that it does not call torch.cuda.device_count before the multiprocessing processes start then it works correctly:

~~~~
import multiprocessing, torch, os

def mp_worker(gpu):
    print(torch.cuda.get_device_properties(gpu))

gpus = list(range(len(os.environ["CUDA_VISIBLE_DEVICES"].split(","))))

processes = [multiprocessing.Process(target=mp_worker, args=(gpui,)) for gpui in gpus]
for process in processes:
    process.start()
for process in processes:
    process.join()
~~~~

Outputs as expected:
~~~~
_CudaDeviceProperties(name='Tesla P100-PCIE-16GB', major=6, minor=0, total_memory=16276MB, multi_processor_count=56)
_CudaDeviceProperties(name='Tesla P100-PCIE-16GB', major=6, minor=0, total_memory=16276MB, multi_processor_count=56)
~~~~

## Environment

~~~~
PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB
GPU 2: Tesla P100-PCIE-16GB
GPU 3: Tesla P100-PCIE-16GB
GPU 4: Tesla P100-PCIE-16GB
GPU 5: Tesla P100-PCIE-16GB
GPU 6: Tesla P100-PCIE-16GB
GPU 7: Tesla P100-PCIE-16GB

Nvidia driver version: 384.145
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy (1.15.4)
[pip3] torch (1.0.0)
[pip3] torchvision (0.2.1)
~~~~

## Additional context

I suspect this has something to do with cuda initialization not working with multiprocessing's fork, but if you wait to initialize cuda within each forked process then it initializes correctly.

### PR DESCRIPTION
**Pull request by ezyang**
This is kind of a bogus solution, but I'm hoping it will work
so I don't have to rewrite all of this code to do it properly.
See comment in code for more details.

Fixes #15734

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/cuda/__init__.py [hunk 1]
```diff
@@ -64,6 +64,26 @@ def is_available():
...
         return False
+    # Lazy initialization here is necessary, as once we
+    # call cudaGetDeviceCount() we have initialized the CUDA
+    # API and forks are no longer permissible.  Calling
+    # _lazy_init() is required so we can give a good error
+    # message in this case.
+    #
+    # However, the way we have written the initialization here
+    # is a bit bogus, as the very act of checking if
+    # the driver is sufficient will initialize the CUDA
```
Comment (soumith) [inline torch/cuda/__init__.py]: no, the act of checking if the driver is sufficient does not initialize the CUDA runtime.
The act of querying the device count initializes the driver.

previously, we used to return the value of a driver check, but not check if `deviceCount > 0`.
Arguably that's a better default than `getDeviceCount > 0` if the result of `getDeviceCount > 0` results in the cudacontext being initialized...

I'm a bit conflicted.

Comment (ezyang) [reply torch/cuda/__init__.py]: For reference, we changed this behavior in 767c96850da0d55a567e91368971ca74a1d302f0

**Comment by ezyang** [comment]
I'm planning to also add a test; @malvika2147 showed me that we can make out-of-process tests pretty easily in the test suite now.

**Comment by ezyang** [comment]
Subsumed by #23030

**Comment by zhangguanheng66** [comment]
Calling torch::utils::set_run_yet_variable_to_false() within THCPModule_getDeviceCount_wrap() seems to resolve the problem. However, Fbcode reports some internal test failures. https://github.com/pytorch/pytorch/pull/23030

**Pull request closed without merging**